src/game.cpp

```diff
@@ -13883,72 +13883,56 @@ void game::spawn_mon(int /*shiftx*/, int /*shifty*/)
     }
 }
 
-// Helper function for game::wait().
-static int convert_wait_chosen_to_turns( int choice ) {
-    switch( choice ) {
-    case 1:
-        return MINUTES( 5 );
-    case 2:
-        return MINUTES( 30 );
-    case 3:
-        return HOURS( 1 );
-    case 4:
-        return HOURS( 2 );
-    case 5:
-        return HOURS( 3 );
-    case 6:
-        return HOURS( 6 );
-    case 7:
-        return calendar::turn.diurnal_time_before( calendar::turn.sunrise() );
-    case 8:
-        return calendar::turn.diurnal_time_before( HOURS( 12 ) );
-    case 9:
-        return calendar::turn.diurnal_time_before( calendar::turn.sunset() );
-    case 10:
-        return calendar::turn.diurnal_time_before( HOURS( 0 ) );
-    case 11:
-    default:
-        return 999999999;
-    }
-}
-
 void game::wait()
 {
-    const bool bHasWatch = u.has_watch();
-
+    std::map<int, int> durations;
     uimenu as_m;
-    as_m.text = _("Wait for how long?");
-    as_m.return_invalid = true;
-    as_m.entries.push_back(uimenu_entry(1, true, '1',
-                                        (bHasWatch) ? _("5 Minutes") : _("Wait 300 heartbeats")));
-    as_m.entries.push_back(uimenu_entry(2, true, '2',
-                                        (bHasWatch) ? _("30 Minutes") : _("Wait 1800 heartbeats")));
 
-    if (bHasWatch) {
-        as_m.entries.push_back(uimenu_entry(3, true, '3', _("1 hour")));
-        as_m.entries.push_back(uimenu_entry(4, true, '4', _("2 hours")));
-        as_m.entries.push_back(uimenu_entry(5, true, '5', _("3 hours")));
-        as_m.entries.push_back(uimenu_entry(6, true, '6', _("6 hours")));
+    const bool has_watch = u.has_watch();
+    const auto add_menu_item = [ &as_m, &durations, has_watch ]
+        ( int retval, int hotkey, const std::string &caption = "", int duration = calendar::INDEFINITELY_LONG ) {
+
+        const bool enabled = duration == calendar::INDEFINITELY_LONG || duration < DAYS( 1 ) - MINUTES( 30 );
+        std::string text( caption );
+
+        if( enabled && has_watch && duration != calendar::INDEFINITELY_LONG ) {
+            const std::string dur_str( calendar::print_duration( duration ) );
+            text += ( text.empty() ? dur_str : string_format( " (%s)", dur_str.c_str() ) );
+        }
+        as_m.entries.push_back( uimenu_entry( retval, enabled, hotkey, text ) );
```

Comment:
A piece of pre-existed code. The `uimenu` class encourages this kind of code (I mean `entries` most definitely should be encapsulated, otherwise it's a design flaw). The replacement is reasonable though.